### DISCUSSION TITLE
Extending BaseUser database model

### DISCUSSION THREAD

**Discussion by daanbeverdam** [fastapi-users/fastapi-users#411] [Q&A] [answered]
Hi @frankie567, first of all thanks for creating this library. Saves me a ton of time!

In an app I'm building, I'm trying to extend the SQLAlchemy user model. I envision to have two user types (let's call them `UserA` and `UserB`) which both have their own unique attributes. My `models.py` file looks something like this:

```
class User(SQLAlchemyBaseUserTable, Base):  # inherits from declarative base and fastapi users
    """Base user, extended from fastapi_users."""
    __tablename__ = "users"
    type = Column(String)
    __mapper_args__ = {'polymorphic_on': type}


class UserA(User):
    """UserA, inherits from User."""
    __tablename__ = 'users_a'
    __mapper_args__ = {'polymorphic_identity': 'user_a'}
    id = Column(None, ForeignKey('users.id'), primary_key=True)
    unique_attribute_a = Column(String)


class UserB(User):
    """UserB inherits from User."""
    __tablename__ = 'users_b'
    __mapper_args__ = {'polymorphic_identity': 'user_b'}
    id = Column(None, ForeignKey('users.id'), primary_key=True)
    unique_attribute_b = Column(String)
```

This all works fine on a database level: I'm getting two separate tables for the user types and one for the base user. 

However, when I register an user through the standard register route, only the base user is created (makes sense).
Is there a supported way to update the register handler so I can create either a `UserB` or `UserA` based on the data that comes in? I've been going through the source code but can't seem to find any. If there isn't, do you have any idea how to solve this within `fastapi_users`? I'm also open to other design patterns.

**Comment by frankie567** [comment]
Hi there,

I understand what you are trying to do here, but I think it's not a good way to do it. You should avoid splitting the same data over different tables in a database.

What you should do is this:
* Have a table `users` where you store common user data (e-mail, password, etc.)
* Have two other tables, one to store the attributes of "A" users and the other to store the attributes of "B" users. Both would be linked by a foreign key to the `users` table.

This way, you have only one common way to retrieve and authenticate users. The rest of the data is specific to your business logic and don't mix with authentication logic.

**Comment by daanbeverdam** [comment, marked as answer]
Thanks for the reply! I agree with the pattern you describe. This is also what the above code accomplishes, sorry if that wasn't really clear. I decided to keep the default register endpoint as I can't (and probably shouldn't) really change it. 

What I ended up doing is:

- Have users create an account with base attributes like email, password, and the type of user they are (A or B). I can also use the type attribute for security scopes.
- When they login for the first time I present the user with an additional form where they can fill in the details needed for the `UserA` or `UserB` model.
